include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -525,337 +544,790 @@ namespace internal
+        gradients_quad += dim*n_q_points;
+        values_quad += n_q_points;
+        values_dofs += shape_info.dofs_per_component_on_cell;
       }
   }
 
+
+
+  template <bool symmetric_evaluate, int dim, int fe_degree, int n_q_points_1d, int n_components, typename Number>
+  struct FEFaceEvaluationImpl
+  {
+    static
+    void evaluate_in_face (const MatrixFreeFunctions::ShapeInfo<Number>  &data,
+                           Number *values_dofs,
+                           Number *values_quad,
+                           Number *gradients_quad,
+                           Number *scratch_data,
+                           const bool               evaluate_val,
+                           const bool               evaluate_grad,
+                           const unsigned int       subface_index)
+    {
+      const AlignedVector<Number> &val1
+        = symmetric_evaluate ? data.shape_values_eo :
+          (subface_index >= GeometryInfo<dim>::max_children_per_cell ?
+           data.shape_values : data.values_within_subface[subface_index%2]);
+      const AlignedVector<Number> &val2
+        = symmetric_evaluate ? data.shape_values_eo :
+          (subface_index >= GeometryInfo<dim>::max_children_per_cell ?
+           data.shape_values : data.values_within_subface[subface_index/2]);
+
+      const AlignedVector<Number> &grad1
+        = symmetric_evaluate ? data.shape_gradients_eo :
+          (subface_index >= GeometryInfo<dim>::max_children_per_cell ?
+           data.shape_gradients : data.gradients_within_subface[subface_index%2]);
+      const AlignedVector<Number> &grad2
+        = symmetric_evaluate ? data.shape_gradients_eo :
+          (subface_index >= GeometryInfo<dim>::max_children_per_cell ?
+           data.shape_gradients : data.gradients_within_subface[subface_index/2]);
+
+      typedef internal::EvaluatorTensorProduct
+      <symmetric_evaluate ? internal::evaluate_evenodd :internal::evaluate_general,
+      dim-1,fe_degree+1,n_q_points_1d,Number> Eval;
+      typedef internal::EvaluatorTensorProduct
+      <internal::evaluate_general,dim-1,fe_degree+1,n_q_points_1d,
+      Number> EvalGeneric;
+      Eval eval1(val1,grad1,AlignedVector<Number>(),
+                 data.fe_degree+1, data.n_q_points_1d);
+      Eval eval2(val2,grad2,AlignedVector<Number>(),
+                 data.fe_degree+1, data.n_q_points_1d);
+
+      const unsigned int size_deg = fe_degree > -1 ?
+                                    Utilities::fixed_int_power<fe_degree+1,dim-1>::value :
+                                    (dim > 1 ? Utilities::fixed_power<dim-1>(data.fe_degree+1) : 1);
+      const unsigned int temp_size = EvalGeneric::dofs_per_cell > EvalGeneric::n_q_points ?
+                                     EvalGeneric::dofs_per_cell : EvalGeneric::n_q_points;
+      static_assert(temp_size > 0, "temp_size should not be zero");
+
+      Number  temp_data[temp_size < 100 ? temp_size : 1];
```

Comment:
can you use boost::container::small_vector here instead?

Reply:
Fixed (with using the provided `scratch_data` array rather than a separate one).